pkg/controller/nativedaemonset/nativedaemonset_controller.go

```diff
@@ -0,0 +1,267 @@
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Calculate desired updated replicas based on partition, which equals to total pods number - partition
+	desiredUpdatedReplicas, err := daemonsetutil.CalculateDesiredUpdatedReplicas(partitionStr, len(pods))
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	klog.Infof("Processing DaemonSet %s/%s with desired updated replicas: %d", daemon.Namespace, daemon.Name, desiredUpdatedReplicas)
+
+	// Process the pods according to partition requirements
+	result, err := r.processBatch(daemon, pods, desiredUpdatedReplicas)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return result, nil
+}
+
+// analyzePods analyzes the current pod state and determines how many pods need to be deleted
+// Returns: podsToDelete slice, needToDelete count, batchCompleted bool, error
+func (r *ReconcileNativeDaemonSet) analyzePods(pods []*corev1.Pod, updateRevision string, desiredUpdatedReplicas int32, daemon *appsv1.DaemonSet) ([]*corev1.Pod, int32, bool, error) {
+	updatedPods := int32(0)
+	podsToDelete := make([]*corev1.Pod, 0)
+
+	// Count updated pods and identify pods to delete
+	for _, pod := range pods {
+		if util.IsConsistentWithRevision(pod.GetLabels(), updateRevision) {
+			updatedPods++
+			klog.Infof("Pod %s/%s matches update revision %s, counting as updated", pod.Namespace, pod.Name, updateRevision)
+		} else {
+			// Pods without the label or with an old revision need to be deleted
+			podsToDelete = append(podsToDelete, pod)
+		}
+	}
+
+	// Check if we already have enough updated pods
+	if updatedPods >= desiredUpdatedReplicas {
+		klog.Infof("Batch completed: have %d updated pods >= %d desired, marking batch as completed", updatedPods, desiredUpdatedReplicas)
+		return podsToDelete, 0, true, nil
+	}
+
+	// Calculate how many pods need to be deleted
+	needToDelete := desiredUpdatedReplicas - updatedPods
+
+	// Get maxUnavailable constraint
+	maxUnavailable := daemonsetutil.GetMaxUnavailable(daemon)
+
+	klog.Infof("Current status - updatedPods: %d, desired: %d, needToDelete: %d, available: %d",
+		updatedPods, desiredUpdatedReplicas, needToDelete, len(podsToDelete))
+
+	// Apply constraints
+	needToDelete = daemonsetutil.ApplyDeletionConstraints(needToDelete, maxUnavailable, len(podsToDelete))
+
+	return podsToDelete, needToDelete, false, nil
+}
+
+// executePodDeletion executes the actual pod deletion based on analysis
+func (r *ReconcileNativeDaemonSet) executePodDeletion(podsToDelete []*corev1.Pod, needToDelete int32, daemon *appsv1.DaemonSet) error {
+	if needToDelete <= 0 {
+		klog.Infof("No pods need to be deleted")
+		return nil
+	}
+
+	maxUnavailable := daemonsetutil.GetMaxUnavailable(daemon)
+	klog.Infof("Planning to delete %d pods (maxUnavailable: %d)", needToDelete, maxUnavailable)
+
+	// Sort pods by creation timestamp to delete oldest first
+	sort.Slice(podsToDelete, func(i, j int) bool {
+		return podsToDelete[i].CreationTimestamp.Before(&podsToDelete[j].CreationTimestamp)
```

Comment:
we should delete new , non-scheduled, not-ready pods first, and we can also considering deletion cost